Detectors/EMCAL/simulation/include/EMCALSimulation/Detector.h

```diff
@@ -64,60 +139,50 @@ class Detector : public o2::Base::Detector
   void ConstructGeometry() override;
 
   ///
+  /// Generate EMCAL envelop (mother volume of all supermodules)
+  ///
+  void CreateEmcalEnvelope();
+
+  ///
   /// Generate tower geometry
   ///
   void CreateShiskebabGeometry();
 
   ///
   /// Generate super module geometry
   ///
-  void CreateSmod(const std::string_view mother = "XEN1");
+  void CreateSupermoduleGeometry(const std::string_view mother = "XEN1");
 
   ///
   /// Generate module geometry (2x2 towers)
   ///
-  void CreateEmod(const std::string_view mother = "SMOD", const std::string_view child = "EMOD");
+  void CreateEmcalModuleGeometry(const std::string_view mother = "SMOD", const std::string_view child = "EMOD");
 
   ///
   /// Generate aluminium plates geometry
   ///
   void CreateAlFrontPlate(const std::string_view mother = "EMOD", const std::string_view child = "ALFP");
 
   ///
-  /// Generate towers in module of 1x1
-  /// Prototype studies, remove?
+  /// Calculate the amount of light seen by the APD for a given track segment (charged particles only)
+  /// Calculation done according to Bricks law
   ///
-  void Trd1Tower1X1(Double_t* parSCM0);
-
-  ///
-  /// Generate towers in module of 3x3
-  /// Prototype studies, remove?
-  ///
-  void Trd1Tower3X3(const Double_t* parSCM0);
-
-  ///
-  /// Used by AliEMCALv0::Trd1Tower3X3
-  /// Prototype studies, remove?
+  /// \param[in] energydeposit Energy deposited by a charged particle in the track segment
+  /// \param[in] tracklength Length of the track segment
+  /// \param[in] charge Track charge (in units of elementary charge)
   ///
-  void PbInTrap(const Double_t parTRAP[11], const std::string_view n);
-
-  ///
-  /// Used by AliEMCALv0::Trd1Tower1X1
-  /// Prototype studies, remove?
-  ///
-  void PbInTrd1(const Double_t* parTrd1, const std::string_view n);
+  Double_t CalculateLightYield(Double_t energydeposit, Double_t tracklength, Int_t charge) const;
 
  private:
   Int_t mBirkC0;
   Double_t mBirkC1;
   Double_t mBirkC2;
 
+  std::set<TrackHit>
```

Comment:
`std::set`s are very expensive, both for both per item overhead and per look up. Depending on your vector size, you might want to use a sorted `std::vector<TrackHit>` and do a binary search or a linear search. For a nice article about binary vs linear search: https://dirtyhandscoding.wordpress.com/2017/08/25/performance-comparison-linear-search-vs-binary-search/. If your vector is very big, you might want to sort a separate index vector and keep `std::vector<TrackHit>` unsorted.

Reply:
Indeed the idea is to use a sorted container, and yes, binary search would be the preferred option as search algorithm. However even a linear search should be more efficient than the current implementation which is performing a linear search on an unsorted list.

My worry for the sorted vector is that I have to sort it completely each time I add a new node so I was wondering whether a std::set should be better suited for this task. However the map structure is very lightweight (integer key and pointer to mapped object) so also the sorting shouldn't be too expensive.

The size of the vector is proportional to the amount of particles passing the EMCAL (one hit per particle - also secondary - per EMCAL module).

Reply:
Yes, you should somehow refactor your algorithm to perform all the insertions first, then sort (once) and then do the lookups. Alternatively, you could keep an array of fixed size array where you bin your hits using a lower precision key. In case there is a conflict you create a new fixed size array and use that for the conflicting keys.
In case there are no conflicts (i.e. you only have unique lower precision keys), you end up with a vector sorted by construction. In case there is conflicts you end up with K sorted vectors where K is the maximum number of conflicts you have for a given shorter key.

Reply:
The purpose is to have a lookup during the stepping I don't see how to perform all insert operations before sorting. As the set is used for search operations it needs to expand in time. As as the amount of entries is a-priori unknown a fixed size array is also not possible.

The second proposal should be nothing else than a hashtable, right? That could work. I think std::unordered_map is the c++ stl implementation for it. I will give it a try.

Reply:
Technically `std::unordered_map` is implemented as a vector of lists, so it's actually slightly different from what I am proposing. Keep in mind that all these structures are optimised for insertion / removal while you most likely want insert once, lookup many in hopefully compact store.

Reply:
What you propose is a hash table (table where access is handled by an index which is obtained  by a hash function of a key - even if you explained it differently). And from what I read std::unordered_map implements a hash table even though the internal implementation is different. And the lookup complexity follows what is to be expected for hashtables.

Indeed lookup operations are expected to be much more abundant then insert / remove operations.

I would prefer sticking to containers which are on the market which show good performance instead of reimplementing the custom solutions. If the std::unordered_map fulfills my needs I see no reason to implement something myself. If you are aware of something better on the market - which is within our software stack - I am open for discussion.

Reply:
Lookup complexity can be the same, but the constants times involved are not the same and neither are per-item overheads. This is real life, not `Algorithms 1` course. ;-) Of course the choice should be done based on a cost - benefit analysis, if you are happy with the `unordered_map`, it's for sure better than an ordered one.

Reply:
We can keep the point in mind and I replace the unordered_map once I have something better. For the moment it is "good enough" and it should be more efficient than the version implemented in AliRoot.